### PR TITLE
Add title screen traffic

### LINKED ISSUE #13: Starter project: title screen should have some traffic
The title screen shows a "screensaver" that bounces randomly around an empty map. It'd be more interesting to have some traffic running, and maybe to follow some agents for a while. https://github.com/dabreegster/abstreet/blob/9ba4f83ca47e0d5902115ac2e471709a7313864f/game/src/pregame.rs#L137

### PR DESCRIPTION
**Pull request by RestitutorOrbis**
This PR is still WIP. Resolves #13.

### CODE REVIEW

File: game/src/pregame.rs [hunk 1]
```diff
@@ -24,8 +27,12 @@ pub struct TitleScreen {
 
 impl TitleScreen {
     pub fn new(ctx: &mut EventCtx, app: &App) -> TitleScreen {
-        let mut rng = app.primary.current_flags.sim_flags.make_rng();
-        TitleScreen {
+       let mut rng = app.primary.current_flags.sim_flags.make_rng();
+       let scenario_generator = ScenarioGenerator::small_run(&app.primary.map);
+       let mut timer = Timer::new("title screen timer");
+       let _scenario = scenario_generator.generate(&app.primary.map, &mut rng, &mut timer);
```
Comment (RestitutorOrbis) [inline game/src/pregame.rs:33]: @dabreegster How do I say that the scenario should run? When I start the game, I don't see any traffic. Furthermore, I  see a  bunch of warnings and I'm not entirely sure why they happen. An excerpt is below:
```
119 warnings:
Can't spawn a Bike ending at border DirectedRoadID(468, forwards); no appropriate lanes there
Can't end_at_border for DirectedRoadID(468, forwards) without a sidewalk
Couldn't fulfill SpawnOverTime { num_agents: 10, start_time: Time(0.0), stop_time: Time(5.0), goal: EndOfRoad(DirectedRoadID { id: RoadID(468), forwards: true }), percent_driving: 0.5, percent_biking: 0.5, percent_use_transit: 0.5 } at all
Can't spawn a Bike ending at border DirectedRoadID(468, forwards); no appropriate lanes there
Can't end_at_border for DirectedRoadID(468, forwards) without a sidewalk
Couldn't fulfill SpawnOverTime { num_agents: 10, start_time: Time(0.0), stop_time: Time(5.0), goal: EndOfRoad(DirectedRoadID { id: RoadID(468), forwards: true }), percent_driving: 0.5, percent_biking: 0.5, percent_use_transit: 0.5 } at all
Can't end_at_border for DirectedRoadID(468, forwards) without a sidewalk
```
There are more warnings but there's a lot of them so I just put a few.

Comment (dabreegster) [reply game/src/pregame.rs:33]: You also need to do:

```
                scenario.instantiate(
                    &mut app.primary.sim,
                    &app.primary.map,
                    &mut rng,
                    timer,
                );
                app.primary
                    .sim
                    .tiny_step(&app.primary.map, &mut app.primary.sim_cb);
```

Comment (RestitutorOrbis) [reply game/src/pregame.rs:33]: Thanks for the suggestion! I got traffic going across now. I see a few options going forward:
- Have the title screen keep bouncing across a world filled with traffic
- Have the camera initially bounce randomly but settle on a random agent that wanders into the camera view
- Have the camera follow a random agent (remove bouncing behaviour entirely)
Which one do you like best?

Comment (dabreegster) [reply game/src/pregame.rs:33]: Awesome! No preference -- what do you think would make a good title screen? I can't imagine most people watch it for long. The second impl may be tricky (you'd have to do something with `AgentCache`, and if it complicates that code just for this case, probably not worth it). If you go with the third, you have to decide what to do after that agent is done.

(And I was wrong, you don't need `tiny_step`, unless you're going to immediately look for an agent to snap to. It just advances the sim 0.1s)

Comment (RestitutorOrbis) [reply game/src/pregame.rs:33]: Hmm, so when I tried removing `tiny_step`, the simulation either isn't running or it's running so slow that it's not visually interesting. With regard to those options, I think I'll abstain from the second option for now and I'll pursue the first and third options. That being said, how would I go about selecting a random agent and detecting when it's trip is done?

Comment (dabreegster) [reply game/src/pregame.rs:33]: I would try mimicking what `SpeedSetting::Fast` does in sandbox/speed. When you get an update event, run the sim forward a little. You can play with the speed -- 5x probably looks a little better.

For picking someone, you could do `sim.get_all_people`. I have a meeting in a moment, but check out info/person and info/trip to map person->trips and see trip status. (Or actually look at `sim.active_agents` maybe)

File: game/src/pregame.rs [hunk 2]
```diff
@@ -15,6 +15,8 @@ use map_model::{Map, PermanentMapEdits};
 use rand::Rng;
 use rand_xorshift::XorShiftRng;
 use std::collections::HashMap;
+use sim::{ScenarioGenerator};
```
Comment (dabreegster) [inline game/src/pregame.rs:18]: style nit: `use sim::ScenarioGenerator;` since there's just one thing

File: game/src/pregame.rs [hunk 3]
```diff
@@ -15,6 +15,8 @@ use map_model::{Map, PermanentMapEdits};
...
+use abstutil::Timer;
```
Comment (dabreegster) [inline game/src/pregame.rs:19]: Please run `cargo +nightly fmt`, this is out of order

File: game/src/pregame.rs [hunk 4]
```diff
@@ -497,15 +511,16 @@ impl Screensaver {
         }
     }
 
-    fn update(&mut self, rng: &mut XorShiftRng, ctx: &mut EventCtx, map: &Map) {
+    fn update(&mut self, rng: &mut XorShiftRng, ctx: &mut EventCtx, app: &mut App) {
         if ctx.input.nonblocking_is_update_event().is_some() {
             ctx.input.use_update_event();
             let dist_along = Duration::realtime_elapsed(self.started) * SPEED;
             if dist_along < self.line.length() {
+                app.primary.sim.tiny_step(&app.primary.map, &mut app.primary.sim_cb);
```
Comment (dabreegster) [inline game/src/pregame.rs:519]: You could move this below the current conditional and remove the `tiny_step` call in `start_bounce`; bit simpler

Comment (dabreegster) [inline game/src/pregame.rs:519]: This jumps 0.1 sim seconds roughly 30 times per real second (`UPDATE_FREQUENCY`), so it's equivalent to about 3x speed. But in practice the frame-rate could differ, so it'd be much safer to do `if let Some(dt) = ctx.input.nonblocking_is_update_event()` and then `                app.primary.sim.time_limited_step(
                    &app.primary.map,
                    multiplier * dt,
                    Duration::seconds(0.033),
                    &mut app.primary.sim_cb,
                );`, just like sandbox/speed does. Using `time_limited_step` prevents the title screen from becoming laggy if the sim gets slow. You could tune multiplier however you like; about 3x looks pretty good from the sample run I did.

**Comment by dabreegster** [comment]
Wonderful! Thanks for the PR!

**Pull request merged**